Fixed #32670 -- Allowed GDALRasters to use any GDAL virtual filesystem.

kxtells opened the pull request:
Hello Everybody. Yesterday I was asking in the django-developers mail list, and today it seems I'm opening my first PR.

This PR removes the limit of `django.contrib.gis.GDALRaster` on which kind of paths you can instantiate. Previously this was hard limited to filesystem paths and GDAL in memory `/vsimem/` paths.

GDAL has full support for virtual filesystems, where the path root specifies something about the protocol to access the file.

https://gdal.org/user/virtual_file_systems.html

With this set of changes you can pass `/vsi<prefix>/` paths through `django.contrib.gis.GDALRaster` letting `GDAL` handle how to access those files.

I wonder if it would be better to provide some of those `vsi` as `GDALRaster` "drivers" like the current "MEM" driver, but for the moment this was the least resistance approach to it.

I added a test with `vsizip` as an example, and tried with `vsis3` manually with no problems.

felixxm commented:
@kxtells Thanks for this patch :+1: All new features require [an accepted ticket](https://code.djangoproject.com/newticket).

kxtells commented:
Thanks @felixxm , I just opened https://code.djangoproject.com/ticket/32670 to track/discuss this feature then.

felixxm reviewed (commented):
@kxtells Thanks :+1:

I squashed commits, added release notes, added `versionchanged` annotation, and pushed small edits.

Please adjust docs about ["Using GDAL’s Virtual Filesystem"](https://docs.djangoproject.com/en/3.2/ref/contrib/gis/gdal/#using-gdal-s-virtual-filesystem).

@yellowcap Can you take a look?

yellowcap commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1111,9 +1111,9 @@ blue.
-    representing a raster file.
+    The first parameter can take three forms: a string representing a file path
+    (filesystem or GDAL virtual filesystem), a dictionary with values defining
+    a new raster, or a bytes object representing a raster file.
```
Maybe add an example. The VSI functionality has gained significant traction lately, so adding an example might be relevant.

kxtells replied:
Thanks @yellowcap , as suggested by @felixxm I'll update the docs in [Using GDALs virtual filesystem](https://docs.djangoproject.com/en/3.2/ref/contrib/gis/gdal/#using-gdal-s-virtual-filesystem), so instead of just pointing it out in a small parentheses I think I'll add two subsections there.

* In Memory Raster -> driver=Mem, current documentation.
* Other virtual filesystems -> not specifically supported via driver but just sending the vsi paths. Some examples and maybe a link to the GDAL documentation

yellowcap replied:
Sounds good!

yellowcap commented on django/contrib/gis/gdal/raster/const.py:
```diff
@@ -65,8 +65,11 @@
 }
 
+# GDAL virtual filesystems prefix.
+VSI_FILESYSTEM_PREFIX = '/vsi'
```
I was wondering if we should make a whitelist of allowed vsi filesystems here, instead of this wildcard approach. I spent some time looking for a list of vsi filesystems that GDAL supports, but did not find one. Probably the supported vsi functionality also depends on the specific build of GDAL anyway.

So overall I think this is fine here and we can leave it as proposed. I just wanted to make this comment here for completeness.

felixxm replied:
I was thinking about the same, but without any static list we cannot perform a proper validation. A wildcard approach is fine IMO, we can add a small warning in the docs, e.g.
```
.. warning:

    Rasters with names starting with `/vsi*/` will be treated as rasters from
    the GDAL virtual filesystems. Django doesn't perform any extra validation.
```

kxtells replied:
I believe the supported `vsi*` will depend on GDAL build. For example the documentation itself states that `/vsicurl` needs GDAL built with libcurl support (https://gdal.org/user/virtual_file_systems.html#vsicurl).

yellowcap replied:
Excellent, looks like we are aligned in our assessment! Thanks for the additional insights on this.

yellowcap reviewed (approved):
Looks good - thanks for the contribution!

kxtells commented:
I added some examples in the documentation, but while doing so I came up with a problem regarding `is_vsi_based`, so I added an extra change to handle it.

Mostly, `is_vsi_based` will return `True` for any `/vsi*` GDALRaster (which I think makes sense). But `vsi_buffer` would fail for any non memory Raster, so I added an extra check in the `vsi_buffer` property to return `None` if the raster is not actually in memory, otherwise the call to capi `VSIGetMemFileBuffer` will return an invalid pointer.

felixxm commented on django/contrib/gis/gdal/raster/source.py:
```diff
@@ -217,7 +217,7 @@ def _flush(self):
     @property
     def vsi_buffer(self):
-        if not self.is_vsi_based:
+        if not self.is_vsi_based or not self.name.startswith(VSI_MEM_FILESYSTEM_BASE_PATH):
```
```suggestion
        if not (self.is_vsi_based and self.name.startswith(VSI_MEM_FILESYSTEM_BASE_PATH)):
```
This change should be covered by tests.

felixxm commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1772,6 +1776,13 @@ Key              Default                           Usage
+filesystems to abstract accessing other kind of files, such as compressed,
+encrypted, or remote files.
+
+Using ``/vsimem/`` Virtual Filesystem
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Maybe:
```suggestion
Using memory-based Virtual Filesystem
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

felixxm commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1817,6 +1828,67 @@ Here's how to create a raster and return it as a file in an
+Using other Virtual Filesystems
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Depending on the local build of GDAL other virtual filesystems might be
+supported and can be used by prepending the provided path with the appropriate
+``/vsi*/`` prefix.
```
```suggestion
Depending on the local build of GDAL other virtual filesystems may be
supported. You can use them by prepending the provided path with the
appropriate ``/vsi*/`` prefix.
```

felixxm commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1817,6 +1828,67 @@ Here's how to create a raster and return it as a file in an
+    Rasters with names starting with `/vsi*/` will be treated as rasters from
+    the GDAL virtual filesystems. Django doesn't perform any extra validation.
+
+The following sections provide some examples on how to use ``vsi*``
+filesystems, a full reference of each filesystem, alongside possible GDAL
+configuration options, can be found in `GDAL Virtual Filesystems
+documentation.`__
+
+__ https://gdal.org/user/virtual_file_systems.html
```
I would change this paragraph to a short note and move it before a warning
```
...
appropriate ``/vsi*/`` prefix. See the
`GDAL Virtual Filesystems documentation`_ for more details.

.. warning:

    Rasters with names starting with `/vsi*/` will be treated as rasters from
    the GDAL virtual filesystems. Django doesn't perform any extra validation.

...

.. `GDAL Virtual Filesystems documentation`: https://gdal.org/user/virtual_file_systems.html
```

felixxm commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1817,6 +1828,67 @@ Here's how to create a raster and return it as a file in an
+
+Instead decompressing the file and instantiating the resulting raster, GDAL can
+directly access compressed files using the ``/vsizip/``, ``/vsitar/`` or
+``/vsigz/`` virtual filesystems::
```
```suggestion
``/vsigzip/`` virtual filesystems::
```

felixxm commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1817,6 +1828,67 @@ Here's how to create a raster and return it as a file in an
+^^^^^^^^^^^^^^^^^^
+
+Instead decompressing the file and instantiating the resulting raster, GDAL can
+directly access compressed files using the ``/vsizip/``, ``/vsitar/`` or
```
Comma is missing before 'or'/'and' (see https://en.wikipedia.org/wiki/Serial_comma).
```suggestion
directly access compressed files using the ``/vsizip/``, ``/vsitar/``, or
```

felixxm commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1817,6 +1828,67 @@ Here's how to create a raster and return it as a file in an
+Network rasters
+^^^^^^^^^^^^^^^
+
+GDAL can support online resources and storage providers transparently. As
+long as it's built with such capabilities.
+
+To access a public raster file with no authentication, use ``/vsicurl/``::
+
+   # Read a remote raster by using /vsicurl/ virtual filesystem
+   >>> from django.contrib.gis.gdal import GDALRaster
+   >>> rst = GDALRaster("/vsicurl/http://example.com/raster.tif")
+   # The name starts with /vsicurl. This raster was read from network
+   >>> rst.name
+   '/vsicurl/http://example.com/raster.tif
+
+
+For commercial storage providers the system should be previously configured for
+authentication and possibly other settings. Refer to `GDAL Virtual Filesystems
+documentation.`__ for each virtual filesystem specific options. Once
+configured, access to the rasters is the same as any other remote raster::
+
+   # Access to a raster in Amazon's S3 cloud store
+   >>> from django.contrib.gis.gdal import GDALRaster
+   >>> rst = GDALRaster(f"/vsis3/path/to/raster.tif")
+   # The name starts with /vsicurl. This raster was read from network
+   >>> rst.name
+
+
+__ https://gdal.org/user/virtual_file_systems.html
```
I think it's too details, folks can always check the GDAL docs for more details. I would limit this to few examples, e.g.
```
Network rasters
^^^^^^^^^^^^^^^

GDAL can support online resources and storage providers transparently. As
long as it's built with such capabilities.

To access a public raster file with no authentication, you can use
``/vsicurl/``::

   >>> from django.contrib.gis.gdal import GDALRaster
   >>> rst = GDALRaster('/vsicurl/http://example.com/raster.tif')
   >>> rst.name
   '/vsicurl/http://example.com/raster.tif'

For commercial storage providers (e.g. ``/vsis3``) the system should be
previously configured for authentication and possibly other settings (see
the `GDAL Virtual Filesystems documentation`_ for available options).

.. _`GDAL Virtual Filesystems documentation`: https://gdal.org/user/virtual_file_systems.html
```

felixxm commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1817,6 +1828,67 @@ Here's how to create a raster and return it as a file in an
+   >>> rst = GDALRaster("/vsicurl/http://example.com/raster.tif")
+   # The name starts with /vsicurl. This raster was read from network
+   >>> rst.name
+   '/vsicurl/http://example.com/raster.tif
```
```suggestion
   '/vsicurl/http://example.com/raster.tif'
```

felixxm commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1817,6 +1828,67 @@ Here's how to create a raster and return it as a file in an
+
+   # Read a remote raster by using /vsicurl/ virtual filesystem
+   >>> from django.contrib.gis.gdal import GDALRaster
+   >>> rst = GDALRaster("/vsicurl/http://example.com/raster.tif")
```
```suggestion
   >>> rst = GDALRaster('/vsicurl/http://example.com/raster.tif')
```

felixxm commented on docs/ref/contrib/gis/gdal.txt:
```diff
@@ -1817,6 +1828,67 @@ Here's how to create a raster and return it as a file in an
+
+To access a public raster file with no authentication, use ``/vsicurl/``::
+
+   # Read a remote raster by using /vsicurl/ virtual filesystem
```
I would chop these comments.

felixxm reviewed (commented):
@kxtells Thanks for updates :+1:

kxtells commented:
Thanks for the comments @felixxm , pushed a commit with an extra test and the manual adjustments suggested.

felixxm commented:
@kxtells Thanks :+1: Looks great :star: Welcome aboard :boat:
